File: app/src/main/java/com/asmat/rolando/popularmovies/fragments/MovieGridFragment.java
package com.asmat.rolando.popularmovies.fragments;

import android.content.Context;
import android.content.Intent;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.v4.app.Fragment;
import android.support.v4.app.LoaderManager;
import android.support.v4.content.AsyncTaskLoader;
import android.support.v4.content.Loader;
import android.support.v7.widget.GridLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.LinearLayout;

import com.asmat.rolando.popularmovies.R;
import com.asmat.rolando.popularmovies.activities.MovieDetailActivity;
import com.asmat.rolando.popularmovies.adapters.MoviesGridAdapter;
import com.asmat.rolando.popularmovies.managers.MovieApiManager;
import com.asmat.rolando.popularmovies.models.Movie;
import com.asmat.rolando.popularmovies.models.MovieAdapterOnClickHandler;
import com.asmat.rolando.popularmovies.utilities.NetworkUtils;
import com.asmat.rolando.popularmovies.utilities.ViewUtils;

import java.util.ArrayList;

/**
 * Created by rolandoasmat on 5/29/17.
 */

public class MovieGridFragment extends Fragment implements MovieAdapterOnClickHandler {

    private MoviesGridAdapter mMoviesGridAdapter;
    private Context mContext;
    private LoaderManager.LoaderCallbacks<ArrayList<Movie>> fetchMoviesCallbacks;
    private int typeOfMovies;
    private int page;
    private RecyclerView mMoviesGrid;
    private LinearLayout mNoInternetView;
    private boolean fetchingMovies = false;
    private final String TAG = "RA:MovieGridFragment:";

    @Override
    public void onAttach(Context context) {
        super.onAttach(context);
        Log.v(tag(), "onAttach");
        mContext = context;
    }

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        Log.v(tag(), "onCreate");
    }

    @Nullable
    @Override
    public View onCreateView(LayoutInflater inflater, @Nullable ViewGroup container,
                             @Nullable Bundle savedInstanceState) {
        Log.v(tag(), "onCreateView");
        View rootView = inflater.inflate(R.layout.fragment_movie_grid, container, false);
        mMoviesGrid = (RecyclerView) rootView.findViewById(R.id.rv_movie_grid);
        mNoInternetView = (LinearLayout) rootView.findViewById(R.id.no_internet_layout);
        GridLayoutManager mMoviesGridLayoutManager;
        if(page == 1) {
            int numOfCol = ViewUtils.calculateNumberOfColumns(mContext);
            mMoviesGridLayoutManager = new GridLayoutManager(mContext, numOfCol);
            mMoviesGridAdapter = new MoviesGridAdapter(this);
            mMoviesGrid.setHasFixedSize(true);
            mMoviesGrid.setLayoutManager(mMoviesGridLayoutManager);
            mMoviesGrid.setAdapter(mMoviesGridAdapter);
            setFetchMoviesLoaderCallback();
            fetchMovies();
        } else {
            mMoviesGridLayoutManager = new GridLayoutManager(mContext, ViewUtils.calculateNumberOfColumns(mContext));
            mMoviesGrid.setLayoutManager(mMoviesGridLayoutManager);
            mMoviesGrid.setAdapter(mMoviesGridAdapter);
        }
        mMoviesGrid.addOnScrollListener(createScrollListener(mMoviesGridLayoutManager));
        mMoviesGrid.setNestedScrollingEnabled(false);
        return rootView;
    }

    @Override
    public void onActivityCreated(@Nullable Bundle savedInstanceState) {
        super.onActivityCreated(savedInstanceState);
        Log.v(tag(), "onActivityCreated");
    }

    @Override
    public void onStart() {
        super.onStart();
        Log.v(tag(), "onStart");
    }

    @Override
    public void onResume() {
        super.onResume();
        Log.v(tag(), "onResume");
        if(!NetworkUtils.isOnline(getContext())) {
            // User has no internet
            mMoviesGrid.setVisibility(View.GONE);
            mNoInternetView.setVisibility(View.VISIBLE);
        } else {
            // Internet connection established
            mMoviesGrid.setVisibility(View.VISIBLE);
            mNoInternetView.setVisibility(View.GONE);
        }
    }

    @Override
    public void onPause() {
        super.onPause();
        Log.v(tag(), "onPause");
    }

    @Override
    public void onStop() {
        super.onStop();
        Log.v(tag(), "onStop");
    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();
        Log.v(tag(), "onDestroyView");
    }

    // DOES NOT get called as user swipes right/left on view pager
    // DOES get called on screen rotation
    @Override
    public void onDestroy() {
        super.onDestroy();
        Log.v(tag(), "onDestroy");
    }

    @Override
    public void onDetach() {
        super.onDetach();
        Log.v(tag(), "onDetach");
    }

    public MovieGridFragment() {
        page = 1;
        Log.v(tag(), "MovieGridFragment constructor.");
    }

    public void setTypeOfMovies(int typeOfMovies) {
        this.typeOfMovies = typeOfMovies;
        Log.v(tag(), "setTypeOfMovies");
        Log.v(tag(), "typeOfMovies: "+typeOfMovies);
    }

    private RecyclerView.OnScrollListener createScrollListener(final GridLayoutManager layoutManager) {
        return new RecyclerView.OnScrollListener() {
            @Override
            public void onScrolled(RecyclerView recyclerView, int dx, int dy) {
                if(dy > 0 && !fetchingMovies) { // User is scrolling down
                    int positionOfLastItem = layoutManager.getItemCount()-1;
                    int currentPositionOfLastVisibleItem = layoutManager.findLastVisibleItemPosition();
                    Log.v(tag(), "onScrolled");
                    Log.v(tag(), "positionOfLastItem: "+positionOfLastItem);
                    Log.v(tag(), "currentPositionOfLastVisibleItem: "+currentPositionOfLastVisibleItem);
                    if(currentPositionOfLastVisibleItem >= positionOfLastItem - 5){
                        fetchMovies();
                    }
                }
            }
        };
    }

    private void fetchMovies() {
        Log.v(tag(), "fetchMovies");
        fetchingMovies = true;
        LoaderManager loaderManager = getActivity().getSupportLoaderManager();
        loaderManager.restartLoader(typeOfMovies, null, fetchMoviesCallbacks);
    }

    @Override
    public void onSaveInstanceState(Bundle outState) {
        Log.v(tag(), "onSaveInstanceState");
    }

    @Override
    public void onClick(Movie movie) {
        Log.v(tag(), "onClick");
        Class destinationClass = MovieDetailActivity.class;
        Intent intentToStartDetailActivity = new Intent(mContext, destinationClass);
        intentToStartDetailActivity.putExtra(MovieDetailActivity.INTENT_EXTRA_TAG, movie);
        startActivity(intentToStartDetailActivity);
    }

    private void setFetchMoviesLoaderCallback() {
        Log.v(tag(), "setFetchMoviesLoaderCallback");

        fetchMoviesCallbacks = new LoaderManager.LoaderCallbacks<ArrayList<Movie>>() {

            @Override
            public Loader<ArrayList<Movie>> onCreateLoader(int id, Bundle args) {
                return new AsyncTaskLoader<ArrayList<Movie>>(mContext) {

                    @Override
                    protected void onStartLoading() {
                        Log.v(tag(), "onStartLoading");
                        forceLoad();
                    }

                    @Override
                    public ArrayList<Movie> loadInBackground() {
                        Log.v(tag(), "loadInBackground");
                        try {
                            Log.v(tag(), "typeOfMovies: "+typeOfMovies);
                            Log.v(tag(), "page: "+page);
                            return MovieApiManager.fetchMoviesOfType(typeOfMovies, page);
                        } catch (Exception e) {
                            e.printStackTrace();
                            return null;
                        }
                    }
                };
            }

            @Override
            public void onLoadFinished(Loader<ArrayList<Movie>> loader, ArrayList<Movie> data) {
                Log.v(tag(), "onLoadFinished");
                // TODO hide loader
                if(data == null) {
                    // TODO show error message
                } else {
                    if(page == 1) {
                        Log.v(tag(), "setting movies");
                        mMoviesGridAdapter.setMovies(data);
                    } else {
                        Log.v(tag(), "adding movies");
                        mMoviesGridAdapter.addMovies(data);
                    }
                    Log.v(tag(), "data: "+data.toString());
                    page++;
                    fetchingMovies = false;
                }
                getLoaderManager().destroyLoader(typeOfMovies);
            }

            @Override
            public void onLoaderReset(Loader<ArrayList<Movie>> loader) {
                Log.v(tag(), "onLoaderReset");
            }
        };
    }

    private String tag() {
        return TAG+typeOfMovies;
    }
}

File: app/src/main/java/com/asmat/rolando/popularmovies/fragments/MovieGridFragment.kt
package com.asmat.rolando.popularmovies.fragments

import android.content.Context
import android.content.Intent
import android.os.Bundle
import android.support.v4.app.Fragment
import android.support.v4.app.LoaderManager
import android.support.v4.content.AsyncTaskLoader
import android.support.v4.content.Loader
import android.support.v7.widget.GridLayoutManager
import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Button
import android.widget.LinearLayout

import com.asmat.rolando.popularmovies.R
import com.asmat.rolando.popularmovies.activities.MovieDetailActivity
import com.asmat.rolando.popularmovies.adapters.MoviesGridBaseAdapter
import com.asmat.rolando.popularmovies.database.Movie
import com.asmat.rolando.popularmovies.managers.MoviesRepository
import com.asmat.rolando.popularmovies.models.MovieAdapterOnClickHandler
import com.asmat.rolando.popularmovies.models.RequestType
import com.asmat.rolando.popularmovies.networking.TheMovieDBClient
import com.asmat.rolando.popularmovies.utilities.NetworkUtils
import com.asmat.rolando.popularmovies.utilities.ViewUtils
import io.reactivex.Single

import java.util.ArrayList

class MovieGridFragment : Fragment(), MovieAdapterOnClickHandler, View.OnClickListener {

    private var moviesGridAdapter: MoviesGridBaseAdapter? = null
    private var mContext: Context? = null
    private var fetchMoviesCallback: Single<List<Movie>>? = null
    @RequestType.Def
    private var typeOfMovies: Int = 0
    private var page: Int = 0
    private var mMoviesGrid: RecyclerView? = null
    private var mNoInternetView: LinearLayout? = null
    private var fetchingMovies = false

    override fun onAttach(context: Context?) {
        super.onAttach(context)
        mContext = context
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?,
                              savedInstanceState: Bundle?): View? {
        val rootView = inflater.inflate(R.layout.fragment_movie_grid, container, false)
        mMoviesGrid = rootView.findViewById(R.id.rv_movie_grid)
        mNoInternetView = rootView.findViewById(R.id.no_internet_layout)
        val mMoviesGridLayoutManager: GridLayoutManager
        if (page == 1) {
            val numOfCol = ViewUtils.calculateNumberOfColumns(mContext!!)
            mMoviesGridLayoutManager = GridLayoutManager(mContext, numOfCol)
            moviesGridAdapter = MoviesGridBaseAdapter(this)
            mMoviesGrid?.setHasFixedSize(true)
            mMoviesGrid?.layoutManager = mMoviesGridLayoutManager
            mMoviesGrid?.adapter = moviesGridAdapter
            setFetchMoviesLoaderCallback()
            fetchMovies()
        } else {
            mMoviesGridLayoutManager = GridLayoutManager(mContext, ViewUtils.calculateNumberOfColumns(mContext!!))
            mMoviesGrid!!.layoutManager = mMoviesGridLayoutManager
            mMoviesGrid!!.adapter = moviesGridAdapter
        }
        mMoviesGrid!!.addOnScrollListener(createScrollListener(mMoviesGridLayoutManager))
        mMoviesGrid!!.isNestedScrollingEnabled = false

        val retryButton = rootView.findViewById<Button>(R.id.retry_button)
        retryButton.setOnClickListener(this)

        return rootView
    }

    override fun onResume() {
        super.onResume()
        checkInternet()
    }

    private fun checkInternet() {
        if (!NetworkUtils.isOnline(context!!)) {
            // User has no internet
            mMoviesGrid!!.visibility = View.GONE
            mNoInternetView!!.visibility = View.VISIBLE
        } else {
            // Internet connection established
            mMoviesGrid!!.visibility = View.VISIBLE
            mNoInternetView!!.visibility = View.GONE
        }
    }

    init {
        page = 1
    }

    fun setTypeOfMovies(@RequestType.Def typeOfMovies: Int) {
        this.typeOfMovies = typeOfMovies
    }

    private fun createScrollListener(layoutManager: GridLayoutManager): RecyclerView.OnScrollListener {
        return object : RecyclerView.OnScrollListener() {
            override fun onScrolled(recyclerView: RecyclerView?, dx: Int, dy: Int) {
                if (dy > 0 && !fetchingMovies) { // User is scrolling down
                    val positionOfLastItem = layoutManager.itemCount - 1
                    val currentPositionOfLastVisibleItem = layoutManager.findLastVisibleItemPosition()
                    if (currentPositionOfLastVisibleItem >= positionOfLastItem - 5) {
                        fetchMovies()
                    }
                }
            }
        }
    }

    private fun fetchMovies() {
        fetchingMovies = true
        fetchMoviesCallback?.subscribe({ result ->
            moviesGridAdapter?.addMovies(result)
            page++
            fetchingMovies = false
        },{ error ->
            // TODO handle error
            fetchingMovies = false
        })
    }

    override fun onClick(movie: Movie) {
        val destinationClass = MovieDetailActivity::class.java
        val intentToStartDetailActivity = Intent(mContext, destinationClass)
        intentToStartDetailActivity.putExtra(MovieDetailActivity.INTENT_EXTRA_MOVIE_ID, movie.id)
        startActivity(intentToStartDetailActivity)
    }

    private fun setFetchMoviesLoaderCallback() {
        fetchMoviesCallback = when (typeOfMovies) {
            RequestType.MOST_POPULAR -> TheMovieDBClient.getPopularMovies(page)
            RequestType.TOP_RATED -> TheMovieDBClient.getTopRatedMovies(page)
            RequestType.NOW_PLAYING -> TheMovieDBClient.getNowPlayingMovies(page)
            RequestType.UPCOMING -> TheMovieDBClient.getUpcomingMovies(page)
            else -> null
        }
    }

    override fun onClick(v: View) {
        when (v.id) {
            R.id.retry_button -> {
                checkInternet()
                fetchMovies()
            }
        }
    }
}